[Style] Split CSS custom property representation into CSS and Style parts

weinig opened the pull request:
#### 5574e1a1ec10f41c9d9782106e9528a2aee6eca0
<pre>
[Style] Split CSS custom property representation into CSS and Style parts
<a href="https://bugs.webkit.org/show_bug.cgi?id=294485">https://bugs.webkit.org/show_bug.cgi?id=294485</a>

Reviewed by Darin Adler.

Split the computed custom property representation out of CSSCustomPropertyValue
and into a new type, Style::CustomProperty.

Previously, the same type, CSSCustomPropertyValue, was used for both the specified (CSS)
value and computed (Style) values, but much of the state space was only valid in
one or the other. Now we are explicit and no longer have invalid states.

With the new non-CSSValue representation, an additional function was needed to
convert to CSSValue for CSSOM (both legacy and typed). Doing this using the
standard extractor based mechanisms led to some test progressions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom.html: (updated to match spec requiring computed URL to be absolute)
* Source/WebCore/Headers.cmake
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/CSSRegisteredCustomProperty.h:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/CSSWideKeyword.h: Added.
* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSVariableParser.cpp:
* Source/WebCore/css/parser/CSSVariableParser.h:
* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.h:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Logging.h: Added.
* Source/WebCore/css/values/primitives/CSSSymbol.cpp:
* Source/WebCore/css/values/primitives/CSSSymbol.h:
* Source/WebCore/dom/ConstantPropertyMap.cpp:
* Source/WebCore/dom/ConstantPropertyMap.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleCustomPropertyData.cpp:
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/PropertyCascade.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleCustomProperty.cpp: Added.
* Source/WebCore/style/StyleCustomProperty.h: Added.
* Source/WebCore/style/StyleCustomPropertyRegistry.cpp: Renamed from Source/WebCore/style/CustomPropertyRegistry.cpp.
* Source/WebCore/style/StyleCustomPropertyRegistry.h: Renamed from Source/WebCore/style/CustomPropertyRegistry.h.
* Source/WebCore/style/StyleExtractor.cpp:
* Source/WebCore/style/StyleExtractor.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolation.cpp:
* Source/WebCore/style/StyleScope.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:

Canonical link: <a href="https://commits.webkit.org/296287@main">https://commits.webkit.org/296287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d696fdd27d5fb9647239176686ff124562e0386d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108044 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18128 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113254 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58560 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110007 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28404 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36259 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82029 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110992 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22516 "Found 1 new test failure: fast/filter-image/clipped-filter.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97342 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21928 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15476 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58000 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91868 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15541 "Found 2 new test failures: imported/w3c/web-platform-tests/url/IdnaTestV2.window.html imported/w3c/web-platform-tests/url/toascii.window.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116381 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35115 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25858 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91055 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35491 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93620 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90849 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23151 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35750 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13507 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40569 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34756 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38114 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36417 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/BlendingKeyframes.cpp:
```diff
@@ -343,7 +343,7 @@ void BlendingKeyframes::updatePropertiesMetadata(const StyleProperties& properti
             }
         } else if (auto* customPropertyValue = dynamicDowncast<CSSCustomPropertyValue>(cssValue)) {
-            if (customPropertyValue->isInherit())
+            if (auto keyword = customPropertyValue->tryCSSWideKeyword(); keyword && *keyword == CSSWideKeyword::Inherit)
```
```suggestion
            if (customPropertyValue->tryCSSWideKeyword() == CSSWideKeyword::Inherit)
```
I'd think the above code would work given the semantics of `==` for `std::optional`.

darinadler commented on Source/WebCore/css/CSSCustomPropertyValue.cpp:
```diff
@@ -46,79 +38,40 @@ Ref<CSSCustomPropertyValue> CSSCustomPropertyValue::createEmpty(const AtomString
+    return WTF::switchOn(m_value,
+        [&](const Ref<CSSVariableReferenceValue>& value) {
+            auto& otherValue = std::get<Ref<CSSVariableReferenceValue>>(other.m_value);
+            return value.ptr() == otherValue.ptr() || value.get() == otherValue.get();
```
Kind of wish there was a function for this idiom, comparing values with an optimization when the pointers are the same.

weinig replied:
It's kinda a version of arePointingToEqualData() that doesn't have the null checks.

darinadler replied:
Like an `arePointingToEqualData` that takes references rather than pointers.

weinig replied:
Yeah, except that `arePointingToEqualData` is abstracted in a way that makes that non-trivial.

The current `arePointingToEqualData` implementation is:

```c++
template<typename T> inline bool arePointingToEqualData(const T& a, const T& b)
{
    return a == b || (a && b && *a == *b);
}
```

So it works with pointers, RefPtr<>, std::unique_ptr<> I guess.

We want some trait that says "this type does not support null". Might be as straightforward as checking if it can be converted to bool.

```c++
template<typename T> inline bool arePointingToEqualData(const T& a, const T& b)
{
    if constexpr (std::convertible_to<T, bool>)
        return a == b || (a && b && *a == *b);
    else
        return a == b || *a == *b;
}
```

That's untested, but might do the trick.

weinig replied:
(do I know why it is marked `inline`, no).

darinadler replied:
I guess one goal would be to make it work with `Ref<>`, doesn’t _really_ matter what technique we use. Could even use overloading.

darinadler replied:
And since `Ref<>` doesn’t support the `*` operator, would either have to add `operator *` or instead overload, or add a version of dereferencing that knows how to be a no-op for types like `Ref<>`, and could add a corresponding version of null checking that knows how to be a no-op for types like `Ref<>`.

darinadler replied:
I suppose "types like `Ref<>`" would include actual references, but then there would be ambiguity because the references could be to an object with pointer-like type!

weinig replied:
I think a set of overloads is probably the best.

I wonder if we can take this one step further and have a type that explicitly opts into this behavior for operator==.

So a reference counted value could either use `Ref<>/RefPtr<>`, where operator== only is true for exactly the same pointer, and `RefWithValueEquality<>/RefPtrWithValueEquality<>`, where operator== uses `arePointingToEqualData`.

The nice thing about having types for this is that when they are members, they can get automatic `operator==(...) = default` for their containing type.

The downsides are generalizing/duplicating lots of Ref/RefPtr related functionality and more cognitive overhead.

In many ways this is the same problem set that Swift looks at with its two equality operators, == and ===.

Will have to noodle on the types part more, but I will add the overload set for arePointingToEqualData in the meantime.

darinadler commented on Source/WebCore/css/CSSCustomPropertyValue.h:
```diff
@@ -126,12 +73,50 @@ class CSSCustomPropertyValue final : public CSSValue {
+    if (auto* keyword = std::get_if<CSSWideKeyword>(&m_value))
+        return *keyword;
+    return { };
+}
```
Can these be in the `.cpp` file and still get sufficiently inlined and collapsed due to link-time optimization, or is it important to put them in the header?

darinadler commented on Source/WebCore/css/parser/CSSPropertyParser.cpp:
```diff
@@ -203,6 +205,21 @@ static RefPtr<CSSPrimitiveValue> consumeCSSWideKeyword(CSSParserTokenRange& rang
+static std::optional<CSSWideKeyword> consumeCSSWideKeyword(CSSParserTokenRange& range)
+{
+    CSSParserTokenRange rangeCopy = range;
+    CSSValueID valueID = rangeCopy.consumeIncludingWhitespace().id();
```
```suggestion
    auto rangeCopy = range;
    auto valueID = rangeCopy.consumeIncludingWhitespace().id();
```

darinadler commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1312,27 +1312,31 @@ inline static bool changedCustomPaintWatchedProperty(const RenderStyle& a, const
+                    if (!valueA)
+                        continue;
+                    if (*valueA != *valueB)
+                        return true;
```
```suggestion
                    if (valueA != valueB && (!valueA || !valueB || *valueA != *valueB))
                        return true;
```
Or we can make a helper function that handles the null checking and write it as just a single function call.

darinadler commented on Source/WebCore/style/StyleBuilder.h:
```diff
@@ -25,17 +25,20 @@
 
 struct CSSRegisteredCustomProperty;
+class CSSCustomPropertyValue;
+enum class CSSWideKeyword : uint8_t;
```
I think we normally do these alphabetical, so `class` before `enum` before `struct`.

darinadler commented on Source/WebCore/style/StyleCustomPropertyRegistry.h:
```diff
@@ -36,6 +36,7 @@ class RenderStyle;
 namespace Style {
 
 class Scope;
+class CustomProperty;
```
Alphabetical please.

weinig replied:
Embarrassing

weinig committed:
[Style] Split CSS custom property representation into CSS and Style parts
https://bugs.webkit.org/show_bug.cgi?id=294485

Reviewed by Darin Adler.

Split the computed custom property representation out of CSSCustomPropertyValue
and into a new type, Style::CustomProperty.

Previously, the same type, CSSCustomPropertyValue, was used for both the specified (CSS)
value and computed (Style) values, but much of the state space was only valid in
one or the other. Now we are explicit and no longer have invalid states.

With the new non-CSSValue representation, an additional function was needed to
convert to CSSValue for CSSOM (both legacy and typed). Doing this using the
standard extractor based mechanisms led to some test progressions.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/typedom.html: (updated to match spec requiring computed URL to be absolute)
* Source/WebCore/Headers.cmake
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/BlendingKeyframes.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/CSSRegisteredCustomProperty.h:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/CSSWideKeyword.h: Added.
* Source/WebCore/css/DOMCSSRegisterCustomProperty.cpp:
* Source/WebCore/css/parser/CSSParserIdioms.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.h:
* Source/WebCore/css/parser/CSSVariableParser.cpp:
* Source/WebCore/css/parser/CSSVariableParser.h:
* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.h:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/MainThreadStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/StylePropertyMapReadOnly.cpp:
* Source/WebCore/css/values/CSSValueAggregates.h:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Logging.h: Added.
* Source/WebCore/css/values/primitives/CSSSymbol.cpp:
* Source/WebCore/css/values/primitives/CSSSymbol.h:
* Source/WebCore/dom/ConstantPropertyMap.cpp:
* Source/WebCore/dom/ConstantPropertyMap.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/StyleCustomPropertyData.cpp:
* Source/WebCore/rendering/style/StyleCustomPropertyData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/MatchedDeclarationsCache.cpp:
* Source/WebCore/style/MatchedDeclarationsCache.h:
* Source/WebCore/style/PropertyCascade.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
* Source/WebCore/style/StyleBuilder.cpp:
* Source/WebCore/style/StyleBuilder.h:
* Source/WebCore/style/StyleCustomProperty.cpp: Added.
* Source/WebCore/style/StyleCustomProperty.h: Added.
* Source/WebCore/style/StyleCustomPropertyRegistry.cpp: Renamed from Source/WebCore/style/CustomPropertyRegistry.cpp.
* Source/WebCore/style/StyleCustomPropertyRegistry.h: Renamed from Source/WebCore/style/CustomPropertyRegistry.h.
* Source/WebCore/style/StyleExtractor.cpp:
* Source/WebCore/style/StyleExtractor.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolation.cpp:
* Source/WebCore/style/StyleScope.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/primitives/StylePosition.cpp:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:

Canonical link: https://commits.webkit.org/296287@main
